hoomd/device.py

```diff
@@ -281,50 +278,89 @@ def enable_profiling(self):
-    """
-    Run simulations on a CPU
+class CPU(_Device):
+    """Select the CPU to execute simulations.
 
     Args:
-        nthreads (int): number of TBB threads
-        communicator (:py:mod:`hoomd.comm.Communicator`): MPI communicator object. Can be left None if using a
-            default MPI communicator
-        msg_file (str): Name of file to write messages to
-        shared_msg_file (str): (MPI only) Name of shared file to write message to (append partition #)
-        notice_level (int): Minimum level of notice messages to print
+        num_threads (int): Number of TBB threads. Set to `None` to auto-select.
+
+        communicator (`hoomd.comm.Communicator`): MPI communicator object.
+            When `None`, create a default communicator that uses all MPI ranks.
+
+        msg_file (str): Filename to write messages to. When `None` use
+            `sys.stdout` and `sys.stderr`.
+
+        shared_msg_file (str): Prefix of filename to write message to (HOOMD
+            will append the MPI partition #). When `None`, messages
+            from all partitions are merged.
+
+        notice_level (int): Minimum level of messages to print.
+
+    .. rubric:: MPI
+
+    In MPI execution environments, create a `CPU` device on every rank.
+
+    .. rubric:: TBB threads
+
+    Set `num_threads` to `None` and TBB will auto-select the number of CPU
+    threads to execute. If the environment variable ``OMP_NUM_THREADS`` is set,
+    HOOMD will use this value. You can also set `num_threads` explicitly.
+
+    Note:
+
+        At this time **very few** features in HOOMD use TBB for threading.
+        Most users should employ MPI for parallel simulations.
     """
 
-    def __init__(self, nthreads=None, communicator=None, msg_file=None, shared_msg_file=None, notice_level=2):
+    def __init__(self,
+                 num_threads=None,
+                 communicator=None,
+                 msg_file=None,
+                 shared_msg_file=None,
+                 notice_level=2):
 
-        _device.__init__(self, communicator, notice_level, msg_file, shared_msg_file)
+        super().__init__(communicator, notice_level, msg_file, shared_msg_file)
 
-        _init_nthreads(nthreads)
+        if num_threads is not None:
+            self.num_threads = num_threads
 
-        self.cpp_exec_conf = _hoomd.ExecutionConfiguration(_hoomd.ExecutionConfiguration.executionMode.CPU,
-                                                           [],
-                                                           False,
-                                                           False,
-                                                           self.comm.cpp_mpi_conf,
-                                                           self.cpp_msg)
+        self.cpp_exec_conf = _hoomd.ExecutionConfiguration(
+            _hoomd.ExecutionConfiguration.executionMode.CPU, [],
+            self.comm.cpp_mpi_conf, self.cpp_msg)
 
+    @property
+    def num_threads(self):
```

Comment:
Is TBB only something that is available in CPU builds? It seems that a user could have both CPU-parallelism and GPU-parallelism happening at the same time during a simulation, and in that case it would make more sense if all the `num_threads` related stuff was in `_Device`.

Reply:
I renamed it to `num_cpu_threads` to avoid confusion when the value is passed to a `GPU` device. This is only needed in extremely narrow corner cases where users run GPU simulations, but use a CPU-only module that is threaded with TBB.